src/Symfony/Component/DependencyInjection/Compiler/CheckTypeHintsPass.php

```diff
@@ -0,0 +1,165 @@
+
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Reference;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+use Symfony\Component\DependencyInjection\Exception\InvalidParameterTypeHintException;
+
+/**
+ * Checks whether injected parameters types are compatible with type hints.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Julien Maulny <jmaulny@darkmira.fr>
+ */
+class CheckTypeHintsPass extends AbstractRecursivePass
+{
+    private $autoload;
+
+    public function __construct(bool $autoload = false)
+    {
+        $this->autoload = $autoload;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function processValue($value, $isRoot = false)
+    {
+        if (!$value instanceof Definition) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        // When service is declared with $container->register(Foo::class); and not $container->register(Foo::class, Foo::class);
+        if (null === $value->getClass()) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        if (!$this->autoload && !\class_exists($this->getClassName($value), false)) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        if (null !== $constructor = $this->getConstructor($value, false)) {
+            $this->checkArgumentsTypeHints($constructor, $value->getArguments());
+        }
+
+        foreach ($value->getMethodCalls() as $methodCall) {
+            $reflectionMethod = $this->getReflectionMethod($value, $methodCall[0]);
+
+            $this->checkArgumentsTypeHints($reflectionMethod /* warn */, $methodCall[1]);
+        }
+
+        return parent::processValue($value, $isRoot);
+    }
+
+    /**
+     * Check type hints for every parameter of a method/constructor.
+     *
+     * @param \ReflectionMethod $reflectionMethod
+     * @param Reference[]       $configurationArguments
+     */
+    private function checkArgumentsTypeHints(\ReflectionMethod $reflectionMethod, array $configurationArguments)
+    {
+        $numberOfRequiredParameters = $reflectionMethod->getNumberOfRequiredParameters();
+
+        if (count($configurationArguments) < $numberOfRequiredParameters) {
+            throw new InvalidArgumentException(sprintf(
+                'In service declaration "%s". Missing parameter in "%s::%s". Expecting %d required parameters, got only %d.',
+                $this->currentId,
+                $reflectionMethod->getDeclaringClass()->getName(),
+                $reflectionMethod->getName(),
+                $numberOfRequiredParameters,
+                count($configurationArguments)
+            ));
+        }
+
+        $reflectionParameters = $reflectionMethod->getParameters();
+        $checksCount = min($reflectionMethod->getNumberOfParameters(), count($configurationArguments));
+
+        if ($reflectionMethod->isVariadic()) {
+            --$checksCount;
```

Comment:
not quite when count($configurationArguments) < NumberOfParameters

Reply:
oops, oh yeah I just understood ! Fixing it

Reply:
Done, covered by this unit test: https://github.com/symfony/symfony/pull/27825/files#diff-5b4f3ed8f2a23401980c3434a7637f47R152